tools/autograd/gen_python_functions.py

```diff
@@ -57,8 +57,35 @@
   ${unpack_self}
   ParsedArgs<${max_args}> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
+
   ${declare_namedtuple_return_types}
   ${dispatch}
+
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+""")
+
+# Check for __torch_function__ overrides
+PY_VARIABLE_FUNCTION_VARARGS = CodeTemplate("""\
```

Comment:
I didn't realize this when I read the original design, but if I understand correctly, you only get `__torch_function__` called if you do an actual function call e.g. `torch.add` rather than a method call `x.add`, is that right? I suppose this makes sense (as method and function names can conflict) but also makes certain use cases (e.g., generically overriding every method with some wrapper code before calling the real implementation) more difficult to do.

Reply:
Yeah. `__torch_function__` is called when we call functions and not when methods are called.

Reply:
The idea is that `torch.somefunction` doesn't know what it's being called on, but a method does know that so there's no need for dispatching to somewhere else.

> (e.g., generically overriding every method with some wrapper code before calling the real implementation) more difficult to do.

That's some form of monkeypatching though that's not necessarily desirable (?). Ideally you'd just want a subclass I think. If those would work as expected, and perhaps we can have a factory function to enable easily wrapping all methods, wouldn't that be better?

If not, we can try to add it here - it sounds a lot like gh-26566.

Reply:
No need to add here, I think it's a good idea to keep the initial implementation focused on the one use case we are interested in.

Reply:
It might be immensely useful to add this to methods as well; see #28361

Reply:
I discussed this with @rgommers offline and it might be best to keep it in a separate PR.

Reply:
Yeah, that's exactly the use case I'm thinking of. Separate PR sgtm.

Reply:
To give you some idea on how I currently monkeypatch in NestedTensor: https://github.com/cpuhrsch/nestedtensor/blob/master/nestedtensor/nested/monkey_patch.py#L32

For methods I indeed just setattr the function (you can re-use a single implementation for e.g. torch.add(input1, input2) for NestedTensor.add(self, input1)). For module-level functions (e.g. torch.add) setattr a new function that scans all inputs for a NestedTensor instance and then dispatches to the new function and otherwise calls into the regular torch function. That means NestedTensor lives above everything else in the dispatch chain.

The main use I'd want to get out of __torch_function__ is to get rid of the module-level overwrites.